[Cuong Ngo] iP

phuccuongngo99 opened the pull request:
DudePro
> "The Brits didn't have DudePro, that's why they lost" - George Washington ([source](https://www.youtube.com/watch?v=xvFZjo5PgG0))

DudePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER FAST* to use

All you need to do is,
1. download it from [here]()
2. double-click it
3. add your tasks

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Reminders (coming soon)

Code for nerds, here is `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

phuccuongngo99 committed:
Implement Level-1 of Duke

phuccuongngo99 committed:
Implement add task and task list

phuccuongngo99 committed:
Add feature mark task as done

phuccuongngo99 committed:
Refractor integer checking function to Util class

phuccuongngo99 committed:
Encapsulate relevant var & func to TaskList class

phuccuongngo99 committed:
Duke: Extract a start command keyword.

phuccuongngo99 committed:
Add ToDos, Events, Deadlines features

phuccuongngo99 committed:
Put all task classes to 1 folder

phuccuongngo99 committed:
Change ACTUAL.txt to ACTUAL.TXT in .gitignore

phuccuongngo99 committed:
Add Automated Text UI Testing

phuccuongngo99 committed:
Refractor relevant codes to TaskManager class

phuccuongngo99 committed:
Add Exceptions if user input is invalid

phuccuongngo99 committed:
Modify TaskManager & Duke to work with exception

phuccuongngo99 committed:
Move all exceptions to 1 folder

phuccuongngo99 committed:
Fix typo in function name in Util class

phuccuongngo99 committed:
Change Duke to only echo exception message

phuccuongngo99 committed:
Remove ACTUAL.TXT and update .gitignore

phuccuongngo99 committed:
Add delete feature

phuccuongngo99 committed:
Fix bug not printing out when 0 task left

phuccuongngo99 committed:
Update input & expected for Level-6 test

phuccuongngo99 committed:
Fix type in EXPECTED.TXT

phuccuongngo99 committed:
Move Util class to utils folder

phuccuongngo99 committed:
Change minor to exceptions folder

phuccuongngo99 committed:
Change minor to tasks folder

phuccuongngo99 committed:
Add commands classes in commands folder

phuccuongngo99 committed:
Change Duke.java to work with new code

Commit:
Add isExit() to Command class

Commit:
Add loading data feature

Commit:
Add support to parse Date but not time

Commit:
Merge branch 'branch-Level-7'

Commit:
Merge branch 'branch-Level-8'

Commit:
Refractor relevant code to Storage class

Commit:
Move storage codes to storage folder

Commit:
Add Ui class

Commit:
Add Ui and Parser class

Commit:
Organize folders into duke package

Commit:
Add JUnit test for Parser & TaskList

Commit:
Minor add in header documentation

Commit:
Add Javadoc to all Command subclasses

Commit:
Add Javadoc to Exceptions subclasses

Commit:
Add class javadoc

Commit:
Fix coding standard

Commit:
Fix code style and add more JavaDoc

Commit:
Implement Find feature by keyword

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-Level-9'

Commit:
Fix bug in find feature

phuccuongngo99 commented:
# DudePro
> "The Brits didn't have DudePro, that was why they lost" - George Washington ([source](https://www.youtube.com/watch?v=xvFZjo5PgG0))

DudePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. Download it from [here]().
2. Double-click it.
3. Add your tasks
4. Let it manage your tasks for you :grin: 

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders (coming soon)

---
Here is a code snippet for you nerds
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

ailing35 commented on src/main/java/duke/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    private String getDescription() throws EmptyDescriptionException {
+        //Get description
```
Is this comment needed? The function seems quite self-explanatory. :)

ailing35 commented on src/main/java/duke/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,78 @@
+import duke.exceptions.DukeException;
+import duke.exceptions.EmptyDescriptionException;
+import duke.exceptions.EmptyTimeException;
+import duke.main.Ui;
+import duke.main.Storage;
+import duke.tasks.Deadline;
+import duke.tasks.Task;
+import duke.main.TaskList;
```
Ordering of imports can be more consistent here.

ailing35 commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,42 @@
+public abstract class Command {
+    protected List<String> userInputList;
+
+    /**
+     * Creates command from userInput in String
+     *
+     * @param userInput User input in string formats
+     * @throws DukeException If user doesn't specify description/time of task
+     */
```
Remember to add punctuations behind each parameter description and the header comments.

ailing35 commented on test/duke/main/TaskListTest.java:
```diff
@@ -0,0 +1,51 @@
+import duke.tasks.Task;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Should this import be listed explicitly instead?

ailing35 reviewed (commented):
Coding standards are well adhered to overall, and I really like the good use of abstraction and OOP. Great job!

Commit:
Merge branch 'add-gradle-support' of github.com:phuccuongngo99/ip

Commit:
Ignore data folder

Commit:
Add Gradle for testing

Commit:
Add checkstyle configuration

Commit:
Merge branch 'branch-A-Gradle'

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
Add GUI to Duke, remove CLI GUI

Commit:
Merge branch 'branch-Level-10'

Commit:
Add Varargs

Commit:
Merge branch 'branch-A-Varargs'

Commit:
Add assertion to code

There's no assertion within the entire code base

There should be assertion to check for illegal variable values entered by users

Let's add assertions in a few place. First, to check if user delete a task
number that's within task list length. Second, to check that storage doesn't
initialize an empty task list on its own.

It's better to let the program fails by assertion than cryptic runtime exception.

Commit:
Change code slightly according to ensure CodeQuality

There are magic values within tasks class and storage class

Let's refractor them into constant class

We should do this to reduce to 1 single source of
value. This will help us easily change the value.

phuccuongngo99 committed:
Merge pull request #1 from phuccuongngo99/branch-A-Assertions

Add assertion to code

phuccuongngo99 committed:
Merge pull request #2 from phuccuongngo99/branch-A-CodeQuality

Change code slightly according to ensure CodeQuality

Commit:
Add CI to run GitHub Actions

Commit:
Change dir of gradle.yml file to correct dir

Commit:
Correct typo .workflows -> workflows

Commit:
Add features to track task done so far

Commit:
Fix bug keep increasing num task done

When user execute done command on already done
task, the number of task done increases

This is clearly a bug since the number of
task done so far shouldn't be updated if
users select done on the done task

Let's add a check within TaskList to only
increment if the task is not done beforehand

It's done this way as number of done task
should be the responsibility of taskList

Commit:
Merge branch 'branch-C-Statistics'

phuccuongngo99 committed:
Minor style change

Commit:
Add Ui and User Guide

Commit:
Merge branch 'master' of github.com:phuccuongngo99/ip

Commit:
Fix all current checkstyle problem

Commit:
Fix docs and checkstyle in test

Commit:
Add tests for Storage class

Commit:
Add feedbacks for other exceptions
